src/provider/systemdmanager/journal_reader.rs

```diff
@@ -0,0 +1,100 @@
+//! This module provides functions for reading from the journal.
+
+use anyhow::{Error, Result};
+use kubelet::log::Sender;
+use std::str;
+use systemd::{journal, journal::JournalRef};
+
+/// Reads journal entries with the given invocation ID and sends the
+/// contained messages.
+///
+/// The options `tail` and `follow` in [`sender`] are taken into account.
+/// If `follow` is `true` then messages are sent until the channel of
+/// [`sender`] is closed. In this case an
+/// [`Err(kubelet::log::SendError::ChannelClosed)`] will be returned.
+pub async fn send_messages(sender: &mut Sender, invocation_id: &str) -> Result<()> {
+    let mut journal = journal::OpenOptions::default().open()?;
+    let journal = journal.match_add("_SYSTEMD_INVOCATION_ID", invocation_id)?;
+
+    if let Some(line_count) = sender.tail() {
+        journal.seek_tail()?;
+        let skipped = journal.previous_skip(line_count as u64 + 1)?;
+        if skipped < line_count + 1 {
+            journal.seek_head()?;
+        }
+
+        if sender.follow() {
+            send_remaining_messages(journal, sender).await?;
+        } else {
+            send_n_messages(journal, sender, line_count).await?;
+        }
+    } else {
+        send_remaining_messages(journal, sender).await?;
+    }
+
+    while sender.follow() {
+        journal.wait(None)?;
+        send_remaining_messages(journal, sender).await?;
+    }
+
+    Ok(())
+}
+
+/// Sends the given number of messages from the journal.
+async fn send_n_messages(
+    journal: &mut JournalRef,
+    sender: &mut Sender,
+    count: usize,
+) -> Result<()> {
+    let mut sent = 0;
+    let mut message_available = true;
+    while sent != count && message_available {
```

Comment:
If `sent` could become greater then `count` then it would be a programming error. If I would use `<` instead of `!=` the programming error would be masked and a potentially incorrect program state would be propagated. If an error occurs further down the line then it would be difficult to find the root cause. Therefore I prefer offensive programming over defensive programming.

If we decide to go the other way, I am also okay.

Actually I am not happy with this code:
* I did not use (somehow) the type system to make the state where `sent > count` unrepresentable.
* I use a counter.
* The reading and sending of messages is intertwined.

But this is probably how it is done in Rust without overcomplicating things,